File: Matlab/1_Robinson_Crusoe/BellmanEq.m
function  value = BellmanEq(kprime)
% Function BellmanEq
%  [value] = BellmanEq(kprime)
%
% 目的:
% k'を一つ与えたときのベルマン方程式を返す関数.
% main_robinson_crusoe.mから呼び出して使う.
%
% グローバル変数: beta gamma alpha capital val_fcn kgrid

global beta gamma alpha capital val_fcn kgrid

%% ベルマン方程式

wealth = capital.^alpha;

cons = wealth - kprime;

% 消費が負値の場合、ペナルティを与えてその値が選ばれないようにする
if cons > 0.0
    util = CRRA(cons, gamma);
else
    util = -10000.0;
end

% 次期の価値関数を線形補間
%vnext = interp1(kgrid, val_fcn, kprime, 'linear', 'extrap');

% 次期の価値関数をスプライン補間
vnext = interp1(kgrid, val_fcn, kprime, 'spline');

value = util + beta.*vnext;

%% トリック(1): k'は正の値しか取らないので、ペナルティを与えてその値が選ばれないようにする
if kprime < 0
    value = -1000000.0;
end

%% トリック(2): "最小化"をするので符号を反転
value = -1*value;
 
return

File: Python/3_6_2_DP_Optimization/BellmanEq.py
def BellmanEq(m,capital,kprime,v_interp):
    """
    -----------------------------------------------
    === (k,k')を所与とした時にベルマン方程式を返す関数 ===
    -----------------------------------------------
    <input>
    ・m: パラメータ等を格納したコンストラクタ
    ・capital: 今期の資本保有量(k)
    ・kprime: 来期の資本保有量(k')
    ・v_: 補間した来期の価値関数
    (※Matlabコード上ではBellmanEq関数内で補間した価値関数を定義しているが、
    今回は明示的に関数の引数として扱う(capitalも同様))
    <output>
    ・value: (k,k')に対するベルマン方程式
    (※コード上では"最小化問題"を解くので符号を反転させた値を返す)
    """
    from CRRA import CRRA

    alpha, beta = m.alpha, m.beta

    if kprime < 0: #(1): k'は正の値しか取らないので、ペナルティを与えてその値が選ばれないようにする
        
        value = -1000000.0
    
    else:

        wealth = capital ** alpha #今期の資本保有量で実現される生産量
        cons = wealth - kprime #消費量
        util = CRRA(m,cons) #効用水準
        value = util + beta*v_interp(kprime) #u(c)+βV_{t+1}(k')

    value = -1*value #(2): コード上では"最小化問題"を解くので符号を反転させた値を返す

    return value

    

    



    